### PR TITLE
[schema + content] feat: RichContent ID

### PR DESCRIPTION
**Pull request by Barackos**
## Motivation
Ability to uniquely identify content.

## Flow
- Upon editor load, a unique `id` is generated & added to the content.
- In case the editor loads with predefined content, containing `id` - the predefined `id` remains unchanged.
- All callbacks report `contentId` upon triggering (`onEditorLoad`, `onViewerLoad`...)

For successfully generating unique ids, `uuid@3.4.0` dependency is used.

## Changes
- Schema: `id` field added to the root hierarchy (e.g. next to `metadata`).
- Converters:
	- `id` is preserved when converting from/to any of our supported converters.
	- `id` is created upon conversion if doesn't exist
- Callbacks: all callbacks report `contentId` upon triggering
- Libs: entry point for duplicating content with new `id` - [duplicate.ts](https://github.com/wix/ricos/pull/3061/files#diff-16af2c9330a393a4a5bb4b6c73afa750561dc7f5f7b7269e3fd8361843dedb6b)
	- entry: `ricos-content/libs/duplicate`
- Unit Testings

### CODE REVIEW

File: packages/ricos-schema/web/src/main/proto/wix/rich_content/v1/rich_content.proto [hunk 1]
```diff
@@ -150,6 +150,7 @@ message Metadata {
     uint32 version = 1 [(wix.api.min) = 1]; // Schema version
     google.protobuf.Timestamp created_timestamp = 2 [(wix.api.readOnly) = true]; // Time content was created
     google.protobuf.Timestamp updated_timestamp = 3; // Time of latest edit
+    google.protobuf.StringValue uuid = 4;
```
Comment (oded-soffrin) [inline packages/ricos-schema/web/src/main/proto/wix/rich_content/v1/rich_content.proto:153]: Just a thought - why not calling it id? (Uuid is implemention detail)
Cant i store content with id "my-content"?

Comment (Barackos) [reply packages/ricos-schema/web/src/main/proto/wix/rich_content/v1/rich_content.proto:153]: Convinced 😄 changing to id.
Regarding "my-content" - currently you can. I thought about adding an ID validator that enforces our format, but we should discuss whether we want to limit consumers (developers) to modify the id as they wish? (considering that we're open source)

Comment (Barackos) [reply packages/ricos-schema/web/src/main/proto/wix/rich_content/v1/rich_content.proto:153]: @lxgreen

File: packages/editor/web/lib/editorStateConversion.ts [hunk 2]
```diff
@@ -7,12 +7,21 @@ import {
...
   obj.VERSION = version;
   return obj;
 };
 
+const addID = obj => {
+  obj.ID = obj.ID || uuidv4();
+  return obj;
+};
+
+const addViables = (obj, version) => pipe(addVersion(obj, version), addID);
```
Comment (lxgreen) [inline packages/editor/web/lib/editorStateConversion.ts:23]:
```suggestion
const addMetadata = (obj, version) => pipe(addVersion(obj, version), addID);
```

Comment (lxgreen) [reply packages/editor/web/lib/editorStateConversion.ts:23]: fp-ts is useless unless you add type annotations

File: packages/editor/web/test/editorStateConversion.spec.js [hunk 3]
```diff
@@ -40,7 +41,7 @@ describe('ContentState conversion', () => {
     const editorState = EditorState.createWithContent(convertFromRaw(rawCopy));
     // eslint-disable-next-line no-unused-vars
     const { VERSION: __, ...rawData } = convertToRaw(editorState.getCurrentContent());
-    expect(rawData).toEqual(raw);
+    expect(compare(rawData, raw, { ignoredKeys: ['ID'] })).toStrictEqual({});
```
Comment (lxgreen) [inline packages/editor/web/test/editorStateConversion.spec.js:44]:
```suggestion
const { VERSION: __, ID: _, ...rawData } = convertToRaw(editorState.getCurrentContent());
expect(rawData).toEqual(raw);
```

File: packages/ricos-content/web/src/converters/draft/toDraft/toDraft.ts [hunk 4]
```diff
@@ -7,6 +7,7 @@ import { generateId } from '../../generateRandomId';
 import { BlockType, HeaderLevel, TO_DRAFT_LIST_TYPE } from '../consts';
 import { RICOS_NODE_TYPE_TO_DATA_FIELD } from '../../../consts';
 import { DraftBlockType } from 'draft-js';
+import { v4 as uuidv4 } from 'uuid';
```
Comment (lxgreen) [inline packages/ricos-content/web/src/converters/draft/toDraft/toDraft.ts:10]:
```suggestion
import { v4 as uuid } from 'uuid';
```

File: packages/ricos-content/web/src/converters/nodeUtils.ts [hunk 5]
```diff
@@ -111,10 +112,17 @@ export const createDecoration = (
...
 ): Decoration => ({ type, ...data });
 
-export const initializeMetadata = (version?: number): Metadata => ({
+export const initializeMetadata = ({
+  version,
+  id = uuidv4(),
+}: {
+  version?: number;
+  id?: string;
+} = {}): Metadata => ({
```
Comment (lxgreen) [inline packages/ricos-content/web/src/converters/nodeUtils.ts:121]:
```suggestion
export const initializeMetadata = (version?: number, id?: string): Metadata => ({
```

File: packages/ricos-content/web/src/converters/nodeUtils.ts [hunk 6]
```diff
@@ -111,10 +112,17 @@ export const createDecoration = (
...
   version: version || LATEST_VERSION,
   createdTimestamp: new Date(),
   updatedTimestamp: new Date(),
+  id,
```
Comment (lxgreen) [inline packages/ricos-content/web/src/converters/nodeUtils.ts:125]:
```suggestion
  id: id || uuid(),
```

File: packages/ricos-content/web/src/duplicate/duplicate.spec.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
+import { duplicate } from './duplicate';
```
Comment (lxgreen) [inline packages/ricos-content/web/src/duplicate/duplicate.spec.ts:1]: the duplicate API should be moved to `/RicosContentAPI` dir like modify, exract, builder

File: packages/ricos-editor/web/src/utils/enforceContentID.ts [hunk 8]
```diff
@@ -0,0 +1,10 @@
+import { DraftContent } from 'wix-rich-content-common';
+import { v4 as uuidv4 } from 'uuid';
```
Comment (lxgreen) [inline packages/ricos-editor/web/src/utils/enforceContentID.ts:2]:
```suggestion
import { v4 as uuid } from 'uuid';
```

File: packages/editor-common/web/src/index.ts [hunk 9]
```diff
@@ -136,7 +137,7 @@ export { DraftOffsetKey };
 export { elementOverflowWithEditor } from './Utils/overflowUtils';
 export { createEditorStyles } from './Utils/createEditorStyles';
 
-export const emptyDraftContent = {
+export const emptyDraftContent: () => DraftContent = () => ({
```
Comment (lxgreen) [inline packages/editor-common/web/src/index.ts:140]:
```suggestion
export const getEmptyDraftContent: () => DraftContent = () => ({
```

File: packages/tiptap-editor/web/src/content-utils/toTiptap/toTiptap.ts [hunk 10]
```diff
@@ -6,6 +6,7 @@ import {
   isNode,
   isTextNode,
 } from '../utils';
+import { v4 as uuidv4 } from 'uuid';
```
Comment (lxgreen) [inline packages/tiptap-editor/web/src/content-utils/toTiptap/toTiptap.ts:9]:
```suggestion
import { v4 as uuid } from 'uuid';
```

File: packages/tiptap-editor/web/src/content-utils/fromTiptap/fromTiptap.ts [hunk 11]
```diff
@@ -34,7 +34,7 @@ const removeIdFromData = value => {
 const removeDocType = ({ type: _, ...content }: JSONContent) => content;
 
 const addMetadata = ({ attrs, ...content }: JSONContent) => ({
-  metadata: attrs?.metadata || initializeMetadata(),
+  metadata: attrs?.metadata || initializeMetadata({ id: attrs?.metadata?.id }),
```
Comment (lxgreen) [inline packages/tiptap-editor/web/src/content-utils/fromTiptap/fromTiptap.ts:37]: if the 2nd part of `||` is executed -- means `attrs?.metadata` is falsy. thus, there's no way that `attrs?.metadata?.id` exists. 
it is equivalent to `initializeMetadata({ id: false })`

**Review by lxgreen — CHANGES_REQUESTED** [review]
beyond all the renaming and refactoring, we have to make sure we handle **every** content generation way, and ID is being added to DraftContent and RichContent.

File: packages/ricos-content/web/src/converters/nodeUtils.ts (hunk 5 above)
Comment (Barackos) [reply packages/ricos-content/web/src/converters/nodeUtils.ts:121]: We have 2 use cases (`fromDraft` & `fromTiptap`) where we want to provide id only.
Separating the parameters would require calling this function like that:
`initializeMetadata(undefined, id)`. I find it less attractive ;) why 2 parameters and not an object?

File: packages/editor/web/lib/editorStateConversion.ts (hunk 2 above)
Comment (Barackos) [reply packages/editor/web/lib/editorStateConversion.ts]: `obj` in this file can be both `ContentState` and `EditorState`. We might need to remove some non-working code from this file, as it is required before I give `obj` a type

File: packages/editor/web/lib/editorStateConversion.ts [hunk 12]
```diff
@@ -7,20 +7,20 @@ import {
 } from '@wix/draft-js';
 import { COLLAPSIBLE_LIST_TYPE, TABLE_TYPE, SOUND_CLOUD_TYPE, VIDEO_TYPE } from 'ricos-content';
 import { version } from '../package.json';
-import { v4 as uuidv4 } from 'uuid';
+import { v4 as uuid } from 'uuid';
 import { pipe } from 'fp-ts/function';
 
-const addVersion = (obj, version) => {
+const addVersion = (obj, version: string): typeof obj => {
```
Comment (lxgreen) [inline packages/editor/web/lib/editorStateConversion.ts:13]: obj is actually DraftContent?

Comment (Barackos) [reply packages/editor/web/lib/editorStateConversion.ts:13]: `DraftContent` | `EditorState`

**Review by lxgreen — APPROVED** [review]

**Pull request merged**